cloud/pkg/cloudstream/tunnelserver.go

```diff
@@ -17,37 +17,51 @@ limitations under the License.
 package cloudstream
 
 import (
+	"context"
 	"crypto/tls"
 	"crypto/x509"
 	"encoding/pem"
 	"fmt"
 	"net/http"
+	"os"
 	"strings"
 	"sync"
 	"time"
 
 	"github.com/emicklei/go-restful"
 	"github.com/gorilla/websocket"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 	certutil "k8s.io/client-go/util/cert"
 	"k8s.io/klog/v2"
 
 	hubconfig "github.com/kubeedge/kubeedge/cloud/pkg/cloudhub/config"
 	streamconfig "github.com/kubeedge/kubeedge/cloud/pkg/cloudstream/config"
+	"github.com/kubeedge/kubeedge/cloud/pkg/common/client"
 	"github.com/kubeedge/kubeedge/pkg/stream"
 )
 
+const (
+	// The amount of time the tunnelserver should sleep between retrying node status updates
+	retrySleepTime        = 20 * time.Millisecond
+	nodeStatusUpdateRetry = 3
+)
+
 type TunnelServer struct {
 	container *restful.Container
 	upgrader  websocket.Upgrader
 	sync.Mutex
 	sessions   map[string]*Session
 	nodeNameIP sync.Map
+	tunnelPort int
```

Comment:
my question is that this `tunnelPort` is from the configuration? and need to be configured by users themselves?

Reply:
No,  users do not need to configure manually.
tunnelPort will be automatically generated when cloudcore is started at first time and saved in configmap. It will be read directly from configmap the next time cloudcore started.